src/System.Text.Utf8/src/System/Text/Utf8/Utf8Encoder.cs

```diff
@@ -117,6 +117,48 @@ public static bool TryDecodeCodePoint(ByteSpan buffer, out UnicodeCodePoint code
 
             return true;
         }
+
+        // TODO: Name TBD
+        // TODO: optimize?
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryDecodeCodePointBackwards(ByteSpan buffer, out UnicodeCodePoint codePoint, out int encodedBytes)
+        {
+            encodedBytes = 1;
+            ByteSpan it = buffer;
+            // TODO: Should we have something like: ByteSpan.(Slice from the back)
+            for (; encodedBytes <= UnicodeConstants.Utf8MaxCodeUnitsPerCodePoint; encodedBytes++, it = it.Slice(0, it.Length - 1))
+            {
+                if (it.Length == 0)
+                {
+                    codePoint = default(UnicodeCodePoint);
+                    encodedBytes = default(int);
+                    return false;
+                }
+
+                // TODO: Should we have ByteSpan.Last?
+                if (!Utf8CodeUnit.IsSurrogate((Utf8CodeUnit)it[it.Length - 1]))
+                {
+                    break;
+                }
+            }
+
+            if (encodedBytes <= UnicodeConstants.Utf8MaxCodeUnitsPerCodePoint)
```

Comment:
Same here? It feels like these are two pieces of the calculation that can be further broken down.